crates/rlbot/src/lib.rs

```diff
@@ -88,97 +90,117 @@ gen_impl_from_flat_packet!(
         // TODO: make this mess nicer
         macro_rules! p {
-            ($x:ident) => {
-                $x::read_as_root(payload)?.try_into().unwrap()
+            ($e:ident) => {
+                Ok(Self::$e)
+            };
+            ($e:ident, $x:ident) => {
+                Ok(Self::$e($x::read_as_root(payload)?.try_into().unwrap()))
+            };
+            ($($n:literal, $($x:ident),+);+) => {
+                match data_type {
+                    $(
+                        $n => p!(
+                            $($x),+
+                        ),
+                    )+
+                    _ => Err(PacketParseError::InvalidDataType(data_type)),
+                }
             };
-        }
 
-        match data_type {
-            0 => Ok(Self::None),
-            1 => Ok(Self::GamePacket(p!(GamePacketRef))),
-            2 => Ok(Self::FieldInfo(p!(FieldInfoRef))),
-            3 => Ok(Self::StartCommand(p!(StartCommandRef))),
-            4 => Ok(Self::MatchConfiguration(p!(MatchConfigurationRef))),
-            5 => Ok(Self::PlayerInput(p!(PlayerInputRef))),
-            6 => Ok(Self::DesiredGameState(p!(DesiredGameStateRef))),
-            7 => Ok(Self::RenderGroup(p!(RenderGroupRef))),
-            8 => Ok(Self::RemoveRenderGroup(p!(RemoveRenderGroupRef))),
-            9 => Ok(Self::MatchComm(p!(MatchCommRef))),
-            10 => Ok(Self::BallPrediction(p!(BallPredictionRef))),
-            11 => Ok(Self::ConnectionSettings(p!(ConnectionSettingsRef))),
-            12 => Ok(Self::StopCommand(p!(StopCommandRef))),
-            13 => Ok(Self::SetLoadout(p!(SetLoadoutRef))),
-            14 => Ok(Self::InitComplete),
-            15 => Ok(Self::ControllableTeamInfo(p!(ControllableTeamInfoRef))),
-            _ => Err(PacketParseError::InvalidDataType(data_type)),
         }
+
+        p!(
+            0, None;
+            1, GamePacket, GamePacketRef;
+            2, FieldInfo, FieldInfoRef;
+            3, StartCommand, StartCommandRef;
+            4, MatchConfiguration, MatchConfigurationRef;
+            5, PlayerInput, PlayerInputRef;
+            6, DesiredGameState, DesiredGameStateRef;
+            7, RenderGroup, RenderGroupRef;
+            8, RemoveRenderGroup, RemoveRenderGroupRef;
+            9, MatchComm, MatchCommRef;
+            10, BallPrediction, BallPredictionRef;
+            11, ConnectionSettings, ConnectionSettingsRef;
+            12, StopCommand, StopCommandRef;
+            13, SetLoadout, SetLoadoutRef;
+            14, InitComplete;
+            15, ControllableTeamInfo, ControllableTeamInfoRef
+        )
     }
 }
 
+pub struct StartingInfo {
+    pub controllable_team_info: ControllableTeamInfo,
+    pub match_configuration: MatchConfiguration,
+    pub field_info: FieldInfo,
+}
+
 pub struct RLBotConnection {
     stream: TcpStream,
     builder: planus::Builder,
-    recv_buf: [u8; u16::MAX as usize],
+    recv_buf: Vec<u8>,
```

Comment:
why put this on the heap? it just makes it slower and saves at max like 60kb of RAM

Reply:
It doesn't "save ram", both the stack and the heap are "in ram." Second, it's very bad practice to put large contiguous slices of memory on the stack, since the stack is limited in size. Things like this, which can get up to 64kb, should always be on the heap.

Reply:
Also, once memory on the heap is allocated, it isn't slower than the same memory on the stack. The only part about the heap that's slower than the stack is when you're constantly allocating & de-allocating items, which is not happening.

Reply:
1. It would save ram if the vec didn't need to grow, no?
2. I know that the heap isn't slower than the stack when reading/writing but we *are* resizing, which could lead to syscalls which are just slowing things down unnecessarily.
3. We aren't allocating that much data (~65kb per bot) but ig we could benefit from the array being boxed instead of on the stack.

Reply:
(I also know that we aren't necessarily resizing the memory underneath every time resize is called, but it will still happen when necessary)

Reply:
`resize` might be more efficient than you think. `Vec`s have a `length` and `capacity`, so the `Vec` will just end up being the size of the largest packet it gets, so realistically, it won't reallocate after the first `GamePacket`. When the `capacity` is sufficient, the function just changes the `length` of the `Vec` which is just updating a `usize` in memory and doesn't touch the heap.

Reply:
I know how `Vec`s and `Vec::resize` works. I still think its unnecessary to have it be dynamic when we're talking about tens of kilobytes.

Reply:
a boxed slice would be acceptable imo, but the real-world overhead is minimal of using a `Vec` instead. your choice

Reply:
boxed slice please :)